### ISSUE TITLE
A signing key must be specified if the specified JWT is digitally signed ?

### ISSUE THREAD

**Issue by donlingliang** [okta/okta-jwt-verifier-java#41]
@bdemers 
I am following the example to build an `AccessTokenVerifier`.
Here is my code:
```
AccessTokenVerifier jwtVerifier = JwtVerifiers.accessTokenVerifierBuilder()
.setIssuer("https://dev-******.okta.com/oauth2/default")
.setAudience("api://default")
.setConnectionTimeout(Duration.ofSeconds(1))
.setReadTimeout(Duration.ofSeconds(1))
.build();

Jwt jwt = jwtVerifier.decode(accessToken);
```
However, when calling **jwtVerifier** `decode` here throws exception below:
```
java.lang.IllegalArgumentException: A signing key must be specified if the specified JWT is digitally signed.
	at io.jsonwebtoken.lang.Assert.notNull(Assert.java:82) ~[auth-1.0-SNAPSHOT.jar:na]
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:357) ~[auth-1.0-SNAPSHOT.jar:na]
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:513) ~[auth-1.0-SNAPSHOT.jar:na]
	at com.okta.jwt.impl.jjwt.TokenVerifierSupport.decode(TokenVerifierSupport.java:61) ~[auth-1.0-SNAPSHOT.jar:na]
	at com.okta.jwt.impl.jjwt.JjwtAccessTokenVerifier.decode(JjwtAccessTokenVerifier.java:45) ~[auth-1.0-SNAPSHOT.jar:na]
```

https://github.com/okta/okta-jwt-verifier-java/issues/13 and checking the okta configuration in console don't help.

By digging a bit, when code is executing to **line 334 - 357** in `DefaultJwtParser.java`, we already decoded the whole accessToken, `payload` and `claims` contains everything same as a network introspect response.

**However**, in the `RemoteJwkSigningKeyResolver.java`, i don't think the jwsHeader(contains two keys `kid` and `alg`) has a correct match with any keys in `keyMap`.

Any places I can attach the Public Key when building the verifier and my jwkUri is `https://dev-******.okta.com/oauth2/default/v1/keys`.

**Comment by bdemers** [comment]
Can you send me the decoded body of the access token you are using?

**Comment by donlingliang** [comment]
Below is the header from the Access Token:
```
{
  "kid": "vTHuz7oYU74F6IeFbwVprj6luhHDNkDOaouc4uDDQtg",
  "alg": "RS256"
}
```

I used Postman to do a GET request with the same jwkUri `https://dev-******.okta.com/oauth2/default/v1/keys` without header and params, it shows me below:
```
{
    "keys": [
        {
            "kty": "RSA",
            "alg": "RS256",
            "kid": "9PGP8uGahk5K4XTWOhXXAIBu7nsA8wzIuoIUx-Yh6QU",
            "use": "sig",
            "e": "AQAB",
            "n": "v1iqTYuLDPs4-eQIHYiaRKF8NTAQ3V5lfIO2aPfQicRfeiWROpDNCH4__JV4STVbkGtcbFSUQdTtwjmV1SEA4XlqnGtbiPywjv87JKCnY7xRM-NhnVRUx5LFM6J50GsSds-x-wrrSKn4_QGmpJfZfFbJp0SyDzrHxHtMJiLWJuLPFnpZ2rR7TcJy-W4fhanOLefYTlrWSli7xjfOuEIYfPTBaZ7TZ7uofwOo5PBWeSyKZ3oxTb5FDAfay3DxN_aXUMMZzEGzAMpNXhFoTC2E_KiOY9dqLtal5lJOBZSeQ60tosLWHlvl86qVlRQbofqbEgQPa5vX3SLd3DcZF_fkjQ"
        },
        {
            "kty": "RSA",
            "alg": "RS256",
            "kid": "S_J9wzAOu0lY8Z8gnV2hYikn-yzDXzm6CRDi-_JPHds",
            "use": "sig",
            "e": "AQAB",
            "n": "2hN70xUANTf9AHXA_AQZrNrrj83T4asZ1gqLFz-50CkFiuz94aDrIU-pXJORSj1MHM_w7iJw8MIkG_eXYvF0TCkBmYxNAMkxGvg1aL7PE85gcss6U1UC4CDRcApb6S6OcUEnZ3MYrj9YD4HwQq5kbhYovBmoshB4raW6uerDEyYbAZckzS9flG5FCMndzaVKGhHaZG6SjikAnm9gqeKNbUYAMQFBzaqjSewVYBzkExXo-yqpJzPRMne6bROTqfTHrW__jB9sibSKU4uV61v1n7Zew3hQuj0ViYTqWJARFokzeA73Gz8hNdLdZbjKYPC9N0W6C8rFJi7uQ5dsLGEazw"
        }
    ]
}
```

Thanks.

**Comment by bdemers** [comment]
Can you include the body of the access token too?

**Comment by donlingliang** [comment]
Here is the payload
```
{
  "ver": 1,
  "jti": "AT.1lbiNw7e9ZS_9AJixE2OvkDjJyAl1o2Uc8f7D-SvpYg.Pu/rW21VbzoJqXu68ELFwVbRMINBqr0ByUKbhzBhk3U=",
  "iss": "https://dev-234850.okta.com",
  "aud": "https://dev-234850.okta.com",
  "sub": "kingslayer_qa_org__don@gettact.com",
  "iat": 1560273220,
  "exp": 1560276820,
  "cid": "0oaobxodmWxqe0eSZ356",
  "uid": "00uphjad4dixpIcHy356",
  "scp": [
    "openid",
    "offline_access"
  ]
}
```

**Comment by bdemers** [comment]
Thanks!  The problem is the `issuer` of the access token, it does NOT match the issuer you have configured for the `AccessTokenVerifier`.

The type of access token you have (it is missing the `/oauth2/<authorization-server-id>` portion) is NOT meant for local validation, clients need to treat it as opaque and use the introspection endpoint instead of treating it as a JWT.

To get a token that you CAN validate locally, you should use the same issuer as above `https://dev-******.okta.com/oauth2/default`(either in postman, service).

Does that help?

**Comment by donlingliang** [comment]
I see, in this case not only the access token request url needs to be changed to `***/default`, also the url that is used to auth/login should be changed. 
A reference for people have the similar issue `/oauth2/<authorization-server-id>` https://developer.okta.com/docs/guides/customize-authz-server/create-authz-server/

Thanks @bdemers !

**Comment by bdemers** [comment]
@donlingliang do you think the following exception would make this more clear:
https://github.com/okta/okta-jwt-verifier-java/pull/42/files#diff-c754a68b1550658bd56d8a37d1623435R68

While also not forcing the user/developer to assume that is the only cause?

**Comment by donlingliang** [comment]
@bdemers
I am curious if we can do an `isSigned()` check in `DefaultJwtParser` before/after **SignatureException** thrown, so that we can specify the causes into two messages(issuer mismatch/the token somehow is not complete) instead. Otherwise the pr looks good to me.

**Comment by bdemers** [comment]
Thanks @donlingliang!
The `DefaultJwtParser` from jjwt is a generic JWT lib, so we wouldn't be able to add issuer logic in there.  That said we might be able to add it along the way somewhere 🤔

**Comment by donlingliang** [comment]
@bdemers Nevermind, your already included that check actually, in **line 56-58**, `TokenVerifierSupport.java`. 😂

**Comment by fabiomontefuscolo** [comment]
Hi,

I got stuck exactly in this part. But in my case, I copied issuer value from postman and pasted that into Java code. The access token has a kid that doesn't match any of the keys in `/oauth2/v1/keys`. 

Also I tried to [create a Authorization Server](https://developer.okta.com/docs/guides/customize-authz-server/create-authz-server/), but there is no such entry in  _Developer Console > API_  menu. Trying `/oauth2/default/v1/keys` I got HTTP 400.

Any clue?

**Comment by bdemers** [comment]
Hi @fabiomontefuscolo ! 

There are a couple of common issues we see that could cause this.
1. Your Okta Org doesn't not have API Access Management enabled
(all developer accounts have this enabled by default)
If you didn't sign up for a non-developer account, you can contact Okta Support and they can help you out.

2. You are using the "admin" URL and appending `/oauth2/default/...`
  For example, a URL formated like this is wrong: `https://dev-123456-admin.okta.com/oauth2/default/v1/keys`, just remove the `-admin`

  The `-admin` is only used for the Okta Developer Conole.

  If you want to browse directly to the Authorization Server configuration you can navigate to something like:
`https://dev-123456-admin.okta.com/admin/oauth2/as` (updating it for your domain of course)


Keep us posted!

**Comment by fabiomontefuscolo** [comment]
@bdemers Thank you very much. Actually, I was inspecting the Okta SDK using Java debugger to understand why it is failing. So it went well comparing the issuer, that is good. But when it loaded the keys from `https://***.oktapreview.com/oauth2/v1/keys`, the public key corresponding to `kid` from access_token is not present.

**Comment by bdemers** [comment]
Ahh!, you need to use a "custom authorization server",  Okta has two types of Authorization Servers, you need one formatted in `https://dev-123456.okta.com/oauth2/<id>/v1/keys`

Out of the box, Developer Orgs have one named `default`.  If you have an older "okta preview" org you might not already have one.  Head over to `https://***-admin. oktapreview.com/admin/oauth2/as` to create one.  If you cannot view this page, contact Okta Support, and they can help you out: https://support.okta.com/help/s/

**Comment by mfickett** [comment]
Just chiming in that I ran into this too, so documentation / error message clarification would be a big help! Some points that were confusing:

*   The default configuration of prod and dev are different. This is clearly described in this thread and in authorization server docs, but meant authorization servers weren't on my mental checklist of reasons prod might not work despite dev working.
*   Web login integration worked with prod but then API token verification didn't. Until I dug into it I didn't understand why this could be; I assumed if I got a good login/token in my web client, it should validate server-side. This is where more info in the error message could be really helpful. Ideally it would say "you are trying to verify this token as if it's from a custom auth server, but the issuer domain looks like an org token, <link to docs>."
*   It sounds like org tokens and custom auth servers need different server-side code for token verification, org tokens require an API call but auth server tokens can just have a signature verified locally? I don't understand / haven't dug into why this is, but in the interest of having dev and prod as similar as possible I have just opted for auth servers in both cases. Could be an opportunity for the Java API to "Just Work" with the same code in either case if it doesn't already.

But I'll temper this by saying Okta docs have been outstanding generally! Even though it took me a little searching to find the right part, once I found the auth server docs it was clear what I needed to do, so thanks!
